lightning/src/ln/channelmanager.rs

```diff
@@ -6210,18 +6304,49 @@ where
 			return;
 		}
 		if valid_mpp {
-			for htlc in sources.drain(..) {
+			let (source_claim_pairs, pending_claim_ptr_opt) = if sources.len() > 1 {
+				let mut pending_claims = PendingMPPClaim {
+					channels_without_preimage: Vec::new(),
+					channels_with_preimage: Vec::new(),
+				};
+				let (source_claim_pairs, channels_without_preimage) = sources.into_iter().filter_map(|htlc| {
+					if let Some(cp_id) = htlc.prev_hop.counterparty_node_id {
+						let htlc_id = htlc.prev_hop.htlc_id;
+						let chan_id = htlc.prev_hop.channel_id;
+						let chan_outpoint = htlc.prev_hop.outpoint;
+						Some((
+							(htlc, Some((cp_id, chan_id, htlc_id))),
+							(cp_id, chan_outpoint, chan_id, htlc_id),
+						))
+					} else {
+						None
+					}
+				}).unzip::<_, _, Vec<_>, _>();
+				pending_claims.channels_without_preimage = channels_without_preimage;
+				(source_claim_pairs, Some(Arc::new(Mutex::new(pending_claims))))
+			} else {
+				(sources.into_iter().map(|htlc| (htlc, None)).collect(), None)
+			};
+			for (htlc, mpp_claim) in source_claim_pairs {
+				let mut pending_mpp_claim = None;
+				let pending_claim_ptr = pending_claim_ptr_opt.as_ref().map(|pending_claim| {
+					pending_mpp_claim = mpp_claim.map(|(cp_id, chan_id, htlc_id)|
+						(cp_id, chan_id, htlc_id, PendingMPPClaimPointer(Arc::clone(pending_claim)))
+					);
+					RAAMonitorUpdateBlockingAction::ClaimedMPPPayment {
+						pending_claim: PendingMPPClaimPointer(Arc::clone(pending_claim)),
+					}
+				});
 				self.claim_funds_from_hop(
 					htlc.prev_hop, payment_preimage,
 					|_, definitely_duplicate| {
 						debug_assert!(!definitely_duplicate, "We shouldn't claim duplicatively from a payment");
-						Some(MonitorUpdateCompletionAction::PaymentClaimed { payment_hash })
+						(Some(MonitorUpdateCompletionAction::PaymentClaimed { payment_hash, pending_mpp_claim }), pending_claim_ptr)
```

Comment:
I find writing it this way a lot less confusing rather than using the intermediate `source_claim_pairs`, getting rid of the remaining `mut`s and some renames:
```rust
			let pending_mpp_claim_opt = if sources.len() > 1 {
				let channels_without_preimage = sources.iter().filter_map(|htlc| {
					htlc.prev_hop.counterparty_node_id.map(|cp_id| {
						(cp_id, htlc.prev_hop.outpoint, htlc.prev_hop.channel_id, htlc.prev_hop.htlc_id)
					})
				}).collect();
				Some(Arc::new(Mutex::new(PendingMPPClaim {
					channels_without_preimage,
					channels_with_preimage: Vec::new()
				})))
			} else { None };
			for htlc in sources.into_iter() {
				let raa_blocker = pending_mpp_claim_opt.as_ref().map(|pending_claim| {
					RAAMonitorUpdateBlockingAction::ClaimedMPPPayment {
						pending_claim: PendingMPPClaimPointer(Arc::clone(pending_claim)),
					}
				});
				let pending_mpp_claim = pending_mpp_claim_opt.as_ref().map(|pending_claim| {
					(htlc.prev_hop.counterparty_node_id.unwrap(), htlc.prev_hop.channel_id,
					 htlc.prev_hop.htlc_id, PendingMPPClaimPointer(Arc::clone(pending_claim)))
				});
				self.claim_funds_from_hop(
					htlc.prev_hop, payment_preimage,
					|_, definitely_duplicate| {
						debug_assert!(!definitely_duplicate, "We shouldn't claim duplicatively from a payment");
						(Some(MonitorUpdateCompletionAction::PaymentClaimed { payment_hash, pending_mpp_claim }), raa_blocker)
					}
				);
			}
```